src/Entity/SmsMessageResult.php

```diff
@@ -191,28 +195,30 @@ public static function baseFieldDefinitions(EntityTypeInterface $entity_type) {
   }
 
   /**
-   * Converts a plain SMS message result into an SMS message result entity.
-   *
-   * @param \Drupal\sms\Message\SmsMessageResultInterface $sms_result
-   *   A plain SMS message result.
-   *
-   * @return \Drupal\sms\Entity\SmsMessageResultInterface
-   *   An SMS message result entity that can be saved.
+   * {@inheritdoc}
    */
-  public static function convertFromMessageResult(PlainMessageResultInterface $sms_result) {
-    if ($sms_result instanceof static) {
-      return $sms_result;
+  public function preSave(EntityStorageInterface $storage) {
+    // SMS message result cannot be saved without a parent SMS message.
+    if (!$this->getSmsMessage()) {
+      throw new \LogicException('No parent SMS message specified for SMS message result');
     }
+    parent::preSave($storage);
+  }
 
-    $new = static::create();
-    $new
-      ->setCreditsBalance($sms_result->getCreditsBalance())
-      ->setCreditsUsed($sms_result->getCreditsUsed())
-      ->setError($sms_result->getError())
-      ->setErrorMessage($sms_result->getErrorMessage())
-      ->setReports($sms_result->getReports());
-
-    return $new;
+  /**
+   * {@inheritdoc}
+   */
+  public static function postLoad(EntityStorageInterface $storage, array &$entities) {
```

Comment:
So long as the internal `$reports` var represents unsaved reports, no previously saved (cached). Then there shouldnt be recursion?